0033-registry-proxy/README.md

```diff
@@ -0,0 +1,368 @@
+
+The root CA, the server (registry) certificate, and the client (proxy) certificate will all be stored in the cluster in the same format as secrets created with [kubectl create secret tls](https://kubernetes.io/docs/reference/kubectl/generated/kubectl_create/kubectl_create_secret_tls/). Users will need to ensure there are tight controls on these secrets to avoid them being maliciously updated or leaked.
+
+Practical risks:
+- Some distros may disallow this
+  - Kind - works
+  - K3D - works
+  - microk8s - works
+  - talos - works, but needs to make the Zarf namespace privileged
+  - k0s - almost certainly works, but I need to get PVCs working on it to test 100%  
+  - k3s - tbd
+  - RKE2 - tbd
+  - OpenShift - HostNetwork works, but hostPort does not. At least need to run `oc adm policy add-scc-to-user privileged -z default -n zarf`
+- Some CNIs may disallow host network or host port
+  - flannel - works
+  - calico - works
+
+## Design Details
+
+<!--
+This section should contain enough information that the specifics of your
+change are understandable. This may include API specs (though not always
+required) or even code snippets. If there's any ambiguity about HOW your
+proposal will be implemented, this is the place to discuss that.
+-->
+
+Initially the proxying component will be based on an existing container image using the `socat` binary ([Alpine socat](https://hub.docker.com/r/alpine/socat)); this is a small and simple image.
+
+`zarf init` will fail if both `--registry-proxy` and `--registry-url` are used. Similarly, init will fail if any of the `--registry-*-file` flags are not empty and `--registry-proxy` is not true.
+
+Running the injector as a daemonset in the airgap requires an image baked into every node, but there is not a distro agnostic way to verify this. The nodeport injector only needs to find one running pod and it can schedule itself on the same node using the same image.
+It's not valid to assume a multi node cluster has a pod running on each node so the DaemonSet injector cannot use this strategy. Instead, Zarf will use an image from a running pod in the kube-system namespace since some distros bake these images into each node. In cases where every kube-system image is not baked into every node, the init could fail. To get around this, users will use the `--injector-image` flag to give an image reference that they know is either baked into each node or able to be pulled in their cluster. The ideal image to be used here will be the pause image as it is required for any pod to start, so every node in a cluster must have access to it. If their initial deployment failed, the user can run `zarf init --registry-proxy --injector-image=registry.k8s.io/pause:3.10`.
+The long lived injector means that the payload configmaps will no longer be deleted from the cluster during `zarf init --registry-proxy`. This amounts to about an additional 32mb of configmaps stored in the cluster permanently.
```

Comment:
If anyone has ideas for improvement here, I'd love to hear it. IMO this is the biggest challenge for this solution to  become the default someday.

Reply:
On second thought maybe pulling an image from kube-system ends up being alright in a strong enough majority of cases. If we look at something like the [talos](https://www.talos.dev/v1.10/advanced/air-gapped/) or [Microk8s](https://microk8s.io/docs/install-offline#h-3-images) air-gap guides, they instruct their users to make the bootstrap images reachable beforehand. 

It seems pretty unlikely for only one node to be pre-loaded with images, would be interesting if anyone knows a distro does that. Maybe a distro has a certain deployment they always run the control plane so their image loaded process differs between servers and agents?

Reply:
Added design details to describe how we'll choose the image. I feel quite good about this now.